common/tasks/interleaved_weighted_round_robin.go

```diff
@@ -0,0 +1,255 @@
+		iwrrChannels:         []*WeightedChannel{},
+	}
+}
+
+func (s *InterleavedWeightedRoundRobinScheduler) Start() {
+	if !atomic.CompareAndSwapInt32(
+		&s.status,
+		common.DaemonStatusInitialized,
+		common.DaemonStatusStarted,
+	) {
+		return
+	}
+
+	s.processor.Start()
+
+	go s.eventLoop()
+
+	s.logger.Info("interleaved weighted round robin task scheduler started")
+}
+
+func (s *InterleavedWeightedRoundRobinScheduler) Stop() {
+	if !atomic.CompareAndSwapInt32(
+		&s.status,
+		common.DaemonStatusStarted,
+		common.DaemonStatusStopped,
+	) {
+		return
+	}
+
+	close(s.shutdownChan)
+
+	s.processor.Stop()
+
+	s.rescheduleTasks()
+
+	s.logger.Info("interleaved weighted round robin task scheduler stopped")
+}
+
+func (s *InterleavedWeightedRoundRobinScheduler) Submit(
+	task PriorityTask,
+
+) {
+	channel := s.getOrCreateTaskChannel(s.priorityToWeight[task.GetPriority()])
+	channel.Chan() <- task
+	s.notifyDispatcher()
+}
+
+func (s *InterleavedWeightedRoundRobinScheduler) eventLoop() {
+	for {
+		select {
+		case <-s.notifyChan:
+			for s.hasRemainingTasks() {
+				weightedChannels := s.channels()
+				s.dispatchTasks(weightedChannels)
+			}
+
+		case <-s.shutdownChan:
+			return
+		}
+	}
+}
+
+func (s *InterleavedWeightedRoundRobinScheduler) getOrCreateTaskChannel(
+	weight int,
+) *WeightedChannel {
+	s.RLock()
+	channel, ok := s.weightToTaskChannels[weight]
+	if ok {
+		s.RUnlock()
+		return channel
+	}
+	s.RUnlock()
+
+	s.Lock()
+	defer s.Unlock()
+
+	channel, ok = s.weightToTaskChannels[weight]
+	if ok {
+		return channel
+	}
+
+	channel = NewWeightedChannel(weight, WeightedChannelDefaultSize)
+	s.weightToTaskChannels[weight] = channel
+
+	weightedChannels := make(WeightedChannels, 0, len(s.weightToTaskChannels))
+	for _, weightedChan := range s.weightToTaskChannels {
+		weightedChannels = append(weightedChannels, weightedChan)
+	}
+	sort.Sort(weightedChannels)
+
+	iwrrChannels := make([]*WeightedChannel, 0, len(weightedChannels))
+	maxWeight := weightedChannels[len(weightedChannels)-1].Weight()
+	for round := maxWeight - 1; round > -1; round-- {
+		for index := len(weightedChannels) - 1; index > -1 && weightedChannels[index].Weight() > round; index-- {
+			iwrrChannels = append(iwrrChannels, weightedChannels[index])
+		}
+	}
+	s.iwrrChannels = iwrrChannels
+
+	return channel
+}
+
+func (s *InterleavedWeightedRoundRobinScheduler) channels() []*WeightedChannel {
+	s.RLock()
+	defer s.RUnlock()
+
+	return s.iwrrChannels
+}
+
+func (s *InterleavedWeightedRoundRobinScheduler) notifyDispatcher() {
+	if s.isStopped() {
+		s.rescheduleTasks()
+		return
+	}
+
+	select {
+	case s.notifyChan <- struct{}{}:
+	default:
+	}
+}
+
+func (s *InterleavedWeightedRoundRobinScheduler) dispatchTasks(
+	channels []*WeightedChannel,
+) {
+
+LoopDispatch:
+	for _, channel := range channels {
+		select {
+		case task := <-channel.Chan():
+			s.processor.Submit(task)
+
+		case <-s.shutdownChan:
+			return
+
+		default:
+			continue LoopDispatch
+		}
+	}
+}
```

Comment:
I'm not following the logic. This would dispatch one task from each weighted channel. I thought weighted means you prefer tasks from higher weighted channel then other channels?

Reply:
the logic assume that that weight of channels is rarely changed, so based one weight, channels appears multiple times

e.g. priority 0 with weight 3, priority 1 with weight 1
then the `channels` on line 209 is [priority 0 chan, priority 0 chan, priority 0 chan, priority 1 chan]